docs(release): add versioning and releases information (#4917)
ZcashFoundation/zebra: book/src/dev/release-process.md

Zebra version numbers indicate the level of changes that are introduced by the release.
This use of [semantic versioning](https://semver.org/ "Semantic Versioning Specification") helps you understand the potential impact of updating to a new version.

Zebra version numbers have three parts: `major.minor.patch`.
For example, version `3.1.11` indicates major version 3, minor version 1, and patch level 11.

The version number is incremented based on the level of change included in the release.

<div class="alert pre-release">

**NOTE**: <br />
As Zebra is in a `pre-release` state (is unstable and might not satisfy the intended compatibility requirements as denoted by its associated normal version).
The pre-release version is denoted by appending a hyphen and a series of dot separated identifiers immediately following the patch version.

</div>

| Level of change | Details |
|:---             |:---     |
| Major release   | Contains significant new features, and commonly correspond to network upgrades; some technical assistance may be needed during the update. When updating to a major release, you may need to follow the specific upgrade instructions provided in the release notes. |

> | Minor release   | Contains new smaller features. Minor releases are should be fully backward-compatible. No technical assistance is expected during update. |
> | Patch release   | Low risk, bug fix release. No technical assistance is expected during update. |

Reviewer: This is good - can we also talk about new features?

```suggestion
| Minor release   | Contains new smaller features. Minor releases are should be fully backward-compatible. No technical assistance is expected during update.  If you want to use the new features in a minor release, you might need to follow the instructions in the release notes. |
| Patch release   | Low risk, bug fix release. No technical assistance is expected during update. |
```

Our support periods will be extended as we gain experience with supporting Zebra releases.

Older Zebra versions that only support previous network upgrades will never be supported, because they are operating on an unsupported Zcash chain fork.

<a id="updating"></a>

### Supported update paths

You can update to any version of Zebra, provided that the following criteria are met:

* The version you want to update *to* is supported.
* The version you want to update *from* is within one major version of the version you want to
    upgrade to.

See [Keeping Up-to-Date](guide/updating "Updating your projects") for more information about updating your Zebra projects to the most recent version.

<a id="previews"></a>

### Preview releases

We let you preview what's coming by providing Release Candidate \(`rc`\) pre-releases for some major releases:

| Pre-release type  | Details |
|:---               |:---     |
| Beta              | The release that is under active development and testing. The beta release is indicated by a release tag appended with the `-beta` identifier, such as  `8.1.0-beta.0`.      |

> | Release candidate | A release that is feature complete and in final testing. A release candidate is indicated by a release tag appended with the `-rc` identifier, such as version `8.1.0-rc.0`. |

Reviewer: I think I originally said "feature-complete", but the team decided not to say it that way in yesterday's meeting.

```suggestion
| Release candidate | A release for final testing of new features. A release candidate is indicated by a release tag appended with the `-rc` identifier, such as version `8.1.0-rc.0`. |
```

* A major release for each network upgrade, whenever there are breaking changes to Zebra (by API, severe bugs or other kind of upgrades)
* Minor releases for significant new Zebra features or severe bug fixes
* A patch release every few weeks

This cadence of releases gives eager developers access to new features as soon as they are fully developed and pass through our code review and integration testing processes, while maintaining the stability and reliability of the platform for production users that prefer to receive features after they have been validated by Zcash and other developers that use the pre-release builds.

<a id="deprecation"></a>

## Deprecation practices

Sometimes "breaking changes", such as the removal of support for RPCs, APIs, and features, are necessary to:

* add new Zebra features,
* improve Zebra performance or reliability,
* stay current with changing dependencies, or
* implement changes in the \(blockchain\) itself.

To make these transitions as straightforward as possible, we make these commitments to you:

* We work hard to minimize the number of breaking changes and to provide migration tools, when possible

> * We follow the deprecation policy described here, so you have time to update your applications to the latest Zebra binaries, RPCs and APIs

Reviewer: This is great!

Sometimes we need to turn things off as soon as possible, because they are causing very serious problems. So let's give ourselves an exception here.

```suggestion
* We follow the deprecation policy described here, so you have time to update your applications to the latest Zebra binaries, RPCs and APIs
* If a feature has critical security or reliability issues, and we need to remove it as soon as possible, we will explain why at the top of the release notes
```

To help ensure that you have sufficient time and a clear path to update, this is our deprecation policy:

| Deprecation stages | Details |
|:---                |:---     |
| Announcement       | We announce deprecated RPCs and features in the [change log](https://github.com/ZcashFoundation/zebra/blob/main/CHANGELOG.md "Zebra change log"). When we announce a deprecation, we also announce a recommended update path. |
| Deprecation period | When a RPC or a feature is deprecated, it is still present until the next major release. A deprecation can be announced in any release, but the removal of a deprecated RPC or feature happens only in major release. Until a deprecated RPC or feature is removed, it is maintained according to the Tier 1 support policy, meaning that only critical and security issues are fixed. |

> | Rust APIs | The Rust APIs of the Zebra crates are unstable and unsupported. Use the `zebrad` commands or JSON-RPCs to interact with Zebra. |

Reviewer: ```suggestion
| Rust APIs | The Rust APIs of the Zebra crates are currently unstable and unsupported. Use the `zebrad` commands or JSON-RPCs to interact with Zebra. |
```

# Zebra versioning and releases

> This document contains the practices that we follow to provide you with a leading-edge application, balanced and with stability.

Reviewer 2: seems like theres an extra "and" should this be: "provide you with a leading-edge application, balanced with stability." ?

We strive to ensure that future changes are always introduced in a predictable way.
We want everyone who depends on Zebra to know when and how new features are added, and to be well-prepared when obsolete ones are removed.

<a id="versioning"></a>

## Zebra versioning

Zebra version numbers indicate the level of changes that are introduced by the release.
This use of [semantic versioning](https://semver.org/ "Semantic Versioning Specification") helps you understand the potential impact of updating to a new version.

Zebra version numbers have three parts: `major.minor.patch`.
For example, version `3.1.11` indicates major version 3, minor version 1, and patch level 11.

The version number is incremented based on the level of change included in the release.

<div class="alert pre-release">

**NOTE**: <br />
As Zebra is in a `pre-release` state (is unstable and might not satisfy the intended compatibility requirements as denoted by its associated normal version).

> The pre-release version is denoted by appending a hyphen and a series of dot separated identifiers immediately following the patch version.

Reviewer 2: an example could be nice here

</div>

| Level of change | Details |
|:---             |:---     |
| Major release   | Contains significant new features, and commonly correspond to network upgrades; some technical assistance may be needed during the update. When updating to a major release, you may need to follow the specific upgrade instructions provided in the release notes. |

> | Minor release   | Contains new smaller features. Minor releases are should be fully backward-compatible. No technical assistance is expected during update.  If you want to use the new features in a minor release, you might need to follow the instructions in the release notes. |

Reviewer 2: Extra "are"; should read: "Minor releases should be fully backward-compatible." instead

| Patch release   | Low risk, bug fix release. No technical assistance is expected during update. |

<a id="supported-releases"></a>

### Supported Releases

Older Zebra versions are always supported until:
* the next Zebra major, minor or patch release.

Initially, we can only guarantee support for the latest Zebra release. We might be able to provide support for earlier releases, or we might ask you to upgrade to the latest release.

Our support periods will be extended as we gain experience with supporting Zebra releases.

Older Zebra versions that only support previous network upgrades will never be supported, because they are operating on an unsupported Zcash chain fork.

<a id="updating"></a>

### Supported update paths

You can update to any version of Zebra, provided that the following criteria are met:

* The version you want to update *to* is supported.
* The version you want to update *from* is within one major version of the version you want to
    upgrade to.

See [Keeping Up-to-Date](guide/updating "Updating your projects") for more information about updating your Zebra projects to the most recent version.

<a id="previews"></a>

> ### Preview releases

Reviewer 2: I think this answers my comment above about giving an example of pre-release versioning, maybe we can link to here from there instead. e.g for an example see "Preview releases"

We strive to ensure that future changes are always introduced in a predictable way.
We want everyone who depends on Zebra to know when and how new features are added, and to be well-prepared when obsolete ones are removed.

<a id="versioning"></a>

## Zebra versioning

Zebra version numbers indicate the level of changes that are introduced by the release.
This use of [semantic versioning](https://semver.org/ "Semantic Versioning Specification") helps you understand the potential impact of updating to a new version.

Zebra version numbers have three parts: `major.minor.patch`.
For example, version `3.1.11` indicates major version 3, minor version 1, and patch level 11.

The version number is incremented based on the level of change included in the release.

<div class="alert pre-release">

**NOTE**: <br />
As Zebra is in a `pre-release` state (is unstable and might not satisfy the intended compatibility requirements as denoted by its associated normal version).
The pre-release version is denoted by appending a hyphen and a series of dot separated identifiers immediately following the patch version.

</div>

| Level of change | Details |
|:---             |:---     |

> | Major release   | Contains significant new features, and commonly correspond to network upgrades; some technical assistance is expected during the update. When updating to a new major release, you might need to <fill-with-things-the-user-might-need-to-do>. |

Reviewer: We can't predict all the things a user might need to do to upgrade, until we have experience managing upgrades.

```suggestion
| Major release   | Contains significant new features, and commonly correspond to network upgrades; some technical assistance may be needed during the update. When updating to a major release, you may need to follow the specific upgrade instructions provided in the release notes. |
```

> | Minor release   | Contains new smaller features. Minor releases are meant to be fully backward-compatible, but may include breaking changes to public APIs. No technical assistance is expected during update. |

Reviewer: semver.org does not allow breaking changes in minor releases.

```suggestion
| Minor release   | Contains new smaller features. Minor releases are should be fully backward-compatible. No technical assistance is expected during update. |
```

| Patch release   | Low risk, bug fix release. No technical assistance is expected during update. |

> 

Reviewer: We need some experience during the release candidate series before we can decide how long to support releases.

But we don't want people thinking we support every Zebra version forever, so we still need to provide some minimum support periods. (It is much easier to extend supported releases than remove them.)

```suggestion

<a id="supported-releases"></a>

### Supported Releases

Older Zebra versions are always supported until:
* the next Zebra major, minor or patch release.

Initially, we can only guarantee support for the latest Zebra release. We might be able to provide support for earlier releases, or we might ask you to upgrade to the latest release.

Our support periods will be extended as we gain experience with supporting Zebra releases.

Older Zebra versions that only support previous network upgrades will never be supported, because they are operating on an unsupported Zcash chain fork.

```

<a id="updating"></a>

### Supported update paths

You can update to any version of Zebra, provided that the following criteria are met:

* The version you want to update *to* is supported.
* The version you want to update *from* is within one major version of the version you want to
    upgrade to.

See [Keeping Up-to-Date](guide/updating "Updating your projects") for more information about updating your Zebra projects to the most recent version.

<a id="previews"></a>

### Preview releases

> We let you preview what's coming by providing "Next" and Release Candidates \(`rc`\) pre-releases for each major and minor release:

Reviewer: I'm not sure if we can promise to always have release candidates, and "next" is not a standard release tag in the Rust or Zcash ecosystems.

```suggestion
We let you preview what's coming by providing Release Candidate \(`rc`\) pre-releases for some major releases:
```

| Pre-release type  | Details |
|:---               |:---     |

> | Next              | The release that is under active development and testing. The next release is indicated by a release tag appended with the `-next` identifier, such as  `8.1.0-next.0`.      |

Reviewer: "next" is non-standard, we usually call this "alpha" or "beta", depending on the stability guarantees.

| Release candidate | A release that is feature complete and in final testing. A release candidate is indicated by a release tag appended with the `-rc` identifier, such as version `8.1.0-rc.0`. |

> The latest `next` or `rc` pre-release version of the documentation is available at [next.Zebra.io](https://next.Zebra.io).

Reviewer: Can we use a standard name for the pre-release documentation website, ideally based on Rust documentation conventions?

<a id="frequency"></a>

> ## Release frequency

Reviewer: I don't think we should commit to major or minor release cycle frequencies, until we have experience doing those releases during the release candidate series. But it is ok to commit to doing regular patch releases.

For example, we might have required major or minor upgrades between network upgrades, due to severe bugs.

Author: I think the only "_commitment_" being done here is `every few weeks`, which could be any amount of weeks. From my PoV we're anyways doing _releases_ every 2 weeks, and `minor` and `major` releases are not setting a specific time. We could go through this text later and change it with the specific cadence.

I also changed the text a bit to allow more flexibility in the release _types_ as you're proposing [docs(release): allow more flexibility with release cycles](https://github.com/ZcashFoundation/zebra/pull/4917/commits/e3a23b09b120add5f1d60a09c61edf9e24300255)

We work toward a regular schedule of releases, so that you can plan and coordinate your updates with the continuing evolution of Zebra.

<div class="alert is-helpful">

Dates are offered as general guidance and are subject to change.

</div>

In general, expect the following release cycle:

> * A major once a network upgrade is required

Reviewer: ```suggestion
* A major release for each network upgrade, and whenever there are breaking changes to Zebra
```

> * 1-3 minor releases for each major release

Reviewer: ```suggestion
* Minor releases for significant new Zebra features
```

> * A patch release build almost semiweekly

Reviewer: Terms like "semiweekly" are hard for English speakers to understand correctly. And our actual release cycle is 2-4 weeks.

```suggestion
* A patch release every few weeks
```

> This cadence of releases gives eager developers access to new features as soon as they are fully developed and pass through our code review and integration testing processes, while maintaining the stability and reliability of the platform for production users that prefer to receive features after they have been validated by Zcash and other developers that use the pre-release builds.

Reviewer: Does this belong in the pre-release section?

<a id="deprecation"></a>

## Deprecation practices

> Sometimes "breaking changes", such as the removal of support for select APIs and features, are necessary to innovate and stay current with new best practices, changing dependencies, or changes in the \(blockchain\) itself.

Reviewer: We need to mention RPCs as a supported interface.
This sentence also might be easier to read if it's split up.

```suggestion
Sometimes "breaking changes", such as the removal of support for RPCs, APIs, and features, are necessary to:
* add new Zebra features,
* improve Zebra performance or reliability,
* stay current with changing dependencies, or
* implement changes in the \(blockchain\) itself.
```

To make these transitions as straightforward as possible, we make these commitments to you:

* We work hard to minimize the number of breaking changes and to provide migration tools, when possible

> * We follow the deprecation policy described here, so you have time to update your applications to the latest APIs and best practices

Reviewer: ```suggestion
* We follow the deprecation policy described here, so you have time to update your applications to the latest Zebra binaries, RPCs and APIs
```

To help ensure that you have sufficient time and a clear path to update, this is our deprecation policy:

| Deprecation stages | Details |
|:---                |:---     |

> | Announcement       | We announce deprecated APIs and features in the [change log](https://github.com/ZcashFoundation/zebra/blob/main/CHANGELOG.md "Zebra change log"). Deprecated APIs appear in the [documentation](missing-link) with ~~strikethrough~~. When we announce a deprecation, we also announce a recommended update path. |

Reviewer: We don't support any Rust APIs yet.

```suggestion
| Announcement       | We announce deprecated RPCs and features in the [change log](https://github.com/ZcashFoundation/zebra/blob/main/CHANGELOG.md "Zebra change log"). When we announce a deprecation, we also announce a recommended update path. |
```

> | Deprecation period | When an API or a feature is deprecated, it is still present in the next major release. After that, deprecated APIs and features are candidates for removal. A deprecation can be announced in any release, but the removal of a deprecated API or feature happens only in major release. Until a deprecated API or feature is removed, it is maintained according to the Tier 1 support policy, meaning that only critical and security issues are fixed. |

Reviewer: We can't promise that RPCs or APIs will remain available for the next major release.

For example, if a network upgrade or out-of-cycle breaking change removes a feature, the API might also need to be removed.

```suggestion
| Deprecation period | When a RPC or a feature is deprecated, it is still present until the next major release. A deprecation can be announced in any release, but the removal of a deprecated RPC or feature happens only in major release. Until a deprecated RPC or feature is removed, it is maintained according to the Tier 1 support policy, meaning that only critical and security issues are fixed. |
```

> | crate dependencies | We only make crate dependency updates that require changes to your applications in a major release. In minor releases, we update peer dependencies by expanding the supported versions, but we do not require projects to update these dependencies until a future major version. This means that during minor Zebra releases, crate dependency updates within Zebra applications and libraries are optional. |

Reviewer: We don't plan on stabilising or publishing any Rust APIs as part of the upcoming release candidate series.

```suggestion
| Rust APIs | The Rust APIs of the Zebra crates are unstable and unsupported. Use the `zebrad` commands or JSON-RPCs to interact with Zebra. |
```

<a id="process"></a>

> ## Release candidate & release process

Reviewer: Keeping separate release branches is a lot of work, so we decided not to do it. If we find out that they are needed during the release candidate series, or later on, we can update the documentation then.

Please remove this section.

It might be good to link to our release PR template, which has our current release process.

Identify the commit from which the release stabilization branch will be made.
Release stabilization branches are used so that development can proceed
unblocked on the `master` branch during the release candidate testing and
bug-fixing process. By convention, release stabilization branches are named
`version-X.Y.0` where `X` and `Y` are the major and minor versions for the
release.

### Create the release stabilization branch

**\<TBD\>**

### Create the release candidate branch

**\<TBD\>**

### Make a tag for the tip of the release candidate branch

**\<TBD\>**

## Make and deploy deterministic builds

**\<TBD\>**

## Add release notes to GitHub

> ## Go/No-Go Status: ⚠️

Reviewer: We don't use this checklist any more, it is very outdated. Please remove it.

If we need to add extra checks to our release process, they should be added to the release PR template or main branch protection rules.

  * ✅ `zebrad` functionality works on platforms that meet its system requirements
* Testing
  * ⚠️ CI Passes
    * ✅  Unit tests pass reliably
    * ✅  Property tests pass reliably
    * ⚠️ Acceptance tests pass reliably
  * ✅ Each Zebra crate [builds individually](https://github.com/ZcashFoundation/zebra/issues/1364)
* Implementation and Launch
  * ✅ All [release blocker bugs](https://github.com/ZcashFoundation/zebra/issues?q=is%3Aopen+is%3Aissue+milestone%3A%22First+Alpha+Release%22+label%3AC-bug) have been fixed
  * ✅ The list of [known serious issues](https://github.com/ZcashFoundation/zebra#known-issues) is up to date
  * ✅ The Zebra crate versions are up to date
  * ⚠️ Users can access [the documentation to deploy `zebrad` nodes](https://github.com/ZcashFoundation/zebra#getting-started)
* User Experience
  * ✅ Build completes within 40 minutes in Zebra's CI
    * ✅ Unused dependencies have been removed (use `cargo-udeps`)
  * ✅ `zebrad` executes normally
    * ✅ `zebrad`'s default logging works reasonably well in a terminal
    * ✅ panics, error logs, and warning logs are rare on mainnet
    * ✅ known panics, errors and warnings have open tickets

## Post Release Task List

> ### Merge the release stabilization branch

Reviewer: Please remove this section, we don't have separate release branches yet.
